### PR TITLE
Register token when trying to open a channel with an unregistered token

### PR DESCRIPTION
**Pull request by andrevmatos**
This should fix the #706 issue.
The fix is simple, but wanted to cover this usecase with tests, which are the intent of this first commit.

- [x] Create a test for open/close channel, both with registered and unregistered token
- [ ] Fix the code, registering the token if it isn't registered yet

### CODE REVIEW

File: raiden/tests/unit/api/test_python_api.py [hunk 1]
```diff
@@ -156,3 +160,55 @@ def test_api_channel_events(raiden_chain):
...
+    assert channel1 in api1.get_channel_list(
+        partner_address=maker_address,
+        token_address=token_address,
+    )
+
+    api0.close(token_address, taker_address)
+
+    gevent.sleep(1)
+
+    assert all(channel.state in (CHANNEL_STATE_CLOSED, CHANNEL_STATE_SETTLED)
```
Comment (czepluch) [inline raiden/tests/unit/api/test_python_api.py]: Why do you use the `CHANNEL_STATE_SETTLED` here? The state can only be closed here since, you don't call `settle()`.

File: raiden/tests/unit/api/test_python_api.py [hunk 2]
```diff
@@ -156,3 +160,55 @@ def test_api_channel_events(raiden_chain):
...
+               for channel in api0.get_channel_list(
+        partner_address=taker_address,
+        token_address=token_address,
+    ))
+    assert all(channel.state in (CHANNEL_STATE_CLOSED, CHANNEL_STATE_SETTLED)
```
Comment (czepluch) [inline raiden/tests/unit/api/test_python_api.py]: same as above

File: raiden/tests/unit/api/test_python_api.py [hunk 3]
```diff
@@ -156,3 +160,55 @@ def test_api_channel_events(raiden_chain):
...
+
+@pytest.mark.parametrize('blockchain_type', ['geth'])
+@pytest.mark.parametrize('number_of_nodes', [2])
+@pytest.mark.parametrize('number_of_tokens', [1])
+@pytest.mark.parametrize('register_tokens', [True, False])
+@pytest.mark.parametrize('settle_timeout', [600])
+def test_channel_open_close(raiden_network, token_addresses, settle_timeout):
+    app0, app1 = raiden_network
+
+    maker_address = app0.raiden.address
```
Comment (czepluch) [inline raiden/tests/unit/api/test_python_api.py]: This naming confuses me a bit. What is meant by `maker`?

File: raiden/tests/unit/api/test_python_api.py [hunk 4]
```diff
@@ -156,3 +160,55 @@ def test_api_channel_events(raiden_chain):
...
+    taker_address = app1.raiden.address
```
Comment (czepluch) [inline raiden/tests/unit/api/test_python_api.py]: `taker` is even more confusing. You don't "take" or "decline" a channel. If somebody opens a channel with you, there will be a channel. Maybe use `participant0` and `participant1` instead.

File: raiden/tests/unit/api/test_python_api.py [hunk 5]
```diff
@@ -156,3 +160,55 @@ def test_api_channel_events(raiden_chain):
...
+
+    token_address = token_addresses[0]
+
+    api0 = RaidenAPI(app0.raiden)
+    api1 = RaidenAPI(app1.raiden)
+    api0.open(
```
Comment (czepluch) [inline raiden/tests/unit/api/test_python_api.py]: Can you assert prior to this point, that the token address is not registered?

File: raiden/tests/unit/api/test_python_api.py (hunk 3 above)
Comment (andrevmatos) [inline raiden/tests/unit/api/test_python_api.py]: It was just the convention used by the other tests: maker is the node opening the channel, taker the other end. Wanted to keep coherent with the other tests to facilitate understanding.

File: raiden/tests/unit/api/test_python_api.py (hunk 5 above)
Comment (andrevmatos) [inline raiden/tests/unit/api/test_python_api.py]: I thought about that, but I'm using this test also for the case where the tokens are previously registered (`@pytest.mark.parametrize('register_tokens', [True, False])`), so I prefered to not assert it, but could add a double check (assert only if `register_tokens` is `False`)

File: raiden/tests/unit/api/test_python_api.py (hunk 3 above)
Comment (czepluch) [inline raiden/tests/unit/api/test_python_api.py]: That's for a token swap. That's something else. If you look at the first tests in this file, that's not the naming that's used. Only for token swaps.

File: raiden/tests/unit/api/test_python_api.py (hunk 1 above)
Comment (andrevmatos) [inline raiden/tests/unit/api/test_python_api.py]: It was @LefterisJP suggestion, since although I don't call settle, `CHANNEL_STATE_SETTLED` is a following state of `CHANNEL_STATE_CLOSED`, so even if the test is a little bit loose, it doesn't get any edge case on it.

File: raiden/tests/unit/api/test_python_api.py (hunk 5 above)
Comment (czepluch) [inline raiden/tests/unit/api/test_python_api.py]: Okay. I would really like to have this so that it's clear what's happening.

File: raiden/tests/unit/api/test_python_api.py (hunk 3 above)
Comment (andrevmatos) [inline raiden/tests/unit/api/test_python_api.py]: Ok, will try changing these var names to something clearer.

File: raiden/tests/unit/api/test_python_api.py (hunk 1 above)
Comment (LefterisJP) [inline raiden/tests/unit/api/test_python_api.py]: Hmmm you misunderstood. Settled comes after `closed` but only if the `settle_timeout` period has passed and someone has called `settle`. It should not be here.

File: raiden/network/rpc/client.py [hunk 6]
```diff
@@ -631,7 +631,10 @@ def __init__(self, jsonrpc_client, registry_address, startgas=GAS_LIMIT,
 
     def manager_address_by_token(self, token_address):
         """ Return the channel manager address for the given token. """
-        return self.proxy.channelManagerByToken.call(token_address)
+        manager_address_encoded = self.proxy.channelManagerByToken.call(token_address)
+        if not manager_address_encoded or manager_address_encoded == '0x':
+            manager_address_encoded = self.add_token(token_address)
```
Comment (hackaugusto) [inline raiden/network/rpc/client.py]: This is wrong IMO, it's changing the `manager_address_by_token` contact, why should a function that get a contract that does not exist register it?

Comment (LefterisJP) [inline raiden/network/rpc/client.py]: You suggest we add one more function on top of it which would also register it? Or since it's only used in two places and presumably in both places we would need it registered if it does not exists simply rename it to something like: `maybe_register_token_and_get_manager()`

Comment (hackaugusto) [inline raiden/network/rpc/client.py]: This is what I consider proper behavior:

If you try to get a proxy to a non-existing channel manager None should be returned, this implies the token is not registered.
For the programmatic interfaces this result should be propagated, so in the rest API case that should result in a 404.
For the UIs the user should either be asked to register it (GUI), or informed that it is not registered and how to do it (CLI)

Comment (czepluch) [inline raiden/network/rpc/client.py]: That would work for me. Then we just need to add an endpoint for registering tokens.

Comment (andrevmatos) [inline raiden/network/rpc/client.py]: Hmm, I thought that was the idea.. I don't know a lot about the architecture, but this solution came in talks with @LefterisJP and @czepluch . If the expected the token would be already registered, this whole PR is pointless, and it should be rejected (including the test), and I can make some changes just to clarify this new expected behavior (`None` if token isn't registered, maybe some logging), and a new PR adding an endpoint for registering new tokens.

Comment (LefterisJP) [inline raiden/network/rpc/client.py]: After looking at the behaviour more closely what @hackaugusto suggests here:

> For the programmatic interfaces this result should be propagated, so in the rest API case that should result in a 404. For the UIs the user should either be asked to register it (GUI), or informed that it is not registered and how to do it (CLI)

sounds like the correct behaviour to have in this situation

**Comment by andrevmatos** [comment]
In talks with @czepluch , @hackaugusto and @LefterisJP , we followed the approach to only explicitly register tokens on the network. Closing it.
